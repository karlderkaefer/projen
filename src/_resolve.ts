import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
