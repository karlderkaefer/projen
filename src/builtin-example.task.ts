console.log("hello, I am a builtin task");
console.log("second line");
