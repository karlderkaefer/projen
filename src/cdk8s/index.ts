export * from "./auto-discover";
export * from "./cdk8s-app-ts";
export * from "./cdk8s-construct";
export * from "./integration-test";
