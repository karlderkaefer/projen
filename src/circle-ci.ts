import { Component } from './component';
import { Project } from './project';
import { YamlFile } from './yaml';

export interface CircleCiProps {
  readonly orbs?: Record<string, IOrb>;
  readonly description?: string;
  readonly jobs?: string[];
  readonly version: string;
  readonly workflows?: Record<string, IWorkflow>[];
}

export interface IOrb {
  readonly name: string;
  readonly version: string;
  toString(): string;
}

export class Orb implements IOrb {
  readonly name: string;
  readonly version: string;

  constructor(name: string, version: string) {
    this.name = name;
    this.version = version;
  }
  public toString(): string {
    return `${this.name}@${this.version}`;
  }
}

export interface IJob {
  readonly requires: string[];
  readonly context: string[];
  readonly filterBranchOnly: string[];
}

export interface IWorkflow {
  readonly jobs: Record<string, IJob[]>[];
}

function renderCircleCiFile(props?: CircleCiProps): object {
  const orbs : Record<string, string> = {};
  for (const [id, orb] of Object.entries(props?.orbs ?? {})) {
    orbs[id] = `${orb.toString()}`;
  }
  return {
    version: props?.version || '2.1',
    description: props?.description,
    orbs: orbs,
    jobs: props?.jobs,
    workflows: props?.workflows,
  };
}

export class CircleCi extends Component {
  constructor(project: Project, props?: CircleCiProps) {
    super(project);
    new YamlFile(project, 'circle.yml', {
      committed: true,
      readonly: true,
      obj: () => this._synthesizeCircleCi(props),
    });
  }
  _synthesizeCircleCi(props?: CircleCiProps): object {
    return renderCircleCiFile(props);
  }
}

