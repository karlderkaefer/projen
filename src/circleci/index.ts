export * from "./circleci";
export * from "./constant";
