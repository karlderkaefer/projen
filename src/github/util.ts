export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}
