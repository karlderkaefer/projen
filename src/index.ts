export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-app-ts';
export * from './cdk8s-construct';
export * from './cdktf-construct';
export * from './component';
export * from './construct-lib';
export * from './dev-env';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './ini';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './logger';
export * from './license';
export * from './makefile';
export * from './node-package';
export * from './node-project';
export * from './object-file';
export * from './option-hints';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './typescript-config';
export * from './version';
export * from './yaml';
export * from './xmlfile';
export * from './source-code';
export * from './upgrade-dependencies';
export * from './circle-ci';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';
export * as json from './json/index'; // <-- we also have json.ts
export * as java from './java';
export * as python from './python';
export * as typescript from './typescript/index'; // <-- we also have typescript.ts
export * as javascript from './javascript';
export * as release from './release';
export * as git from './git';
