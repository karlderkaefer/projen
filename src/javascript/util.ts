import { basename, dirname, extname, join, sep } from "path";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}
