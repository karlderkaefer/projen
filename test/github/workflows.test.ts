import { Project } from "../../src/project";
import { synthSnapshot, TestProject } from "../util";

test("no workflow", () => {
  // GIVEN
  const p = new TestProject({
    githubOptions: {
      workflows: false,
    },
  });

  // THEN
  const workflows = synthWorkflows(p);
  expect(Object.keys(workflows).length).toEqual(0);
});

test("adding empty workflow", () => {
  // GIVEN
  const p = new TestProject();

  // WHEN
  p.github?.addWorkflow("my-workflow");

  // THEN
  const workflows = synthWorkflows(p);
  expect(workflows[".github/workflows/my-workflow.yml"]).toMatchSnapshot();
});

test("throws when adding workflow with existing name", () => {
  // GIVEN
  const p = new TestProject({
    stale: true,
  });

  // THEN
  expect(() => p.github?.addWorkflow("stale")).toThrow(
    /there is already a file under/
  );
});

test("throws when adding workflow with adding a job with no runners specified", () => {
  // GIVEN
  const p = new TestProject({
    stale: true,
  });
  // WHEN
  const workflow = p.github?.addWorkflow("my-workflow");

  // THEN
  expect(() =>
    workflow?.addJobs({
      job1: {
        permissions: {},
        steps: [],
        runsOn: [],
      },
    })
  ).toThrow(/at least one runner selector labels must be provided/);
});

test("tryFind valid workflow", () => {
  // GIVEN
  const p = new TestProject();

  // WHEN
  p.github?.addWorkflow("workflow1");
  p.github?.addWorkflow("workflow2");
  const workflow1 = p.github?.tryFindWorkflow("workflow1");

  // THEN
  const workflows = synthWorkflows(p);
  expect(workflows[".github/workflows/workflow1.yml"]).toBeDefined();
  expect(workflows[".github/workflows/workflow2.yml"]).toBeDefined();
  expect(workflow1).toBeDefined();
  expect(workflow1?.name).toEqual("workflow1");
});

test("tryFind unknown workflow", () => {
  // GIVEN
  const p = new TestProject();

  // WHEN
  p.github?.addWorkflow("workflow1");
  p.github?.addWorkflow("workflow2");
  const workflow3 = p.github?.tryFindWorkflow("workflow3");

  // THEN
  const workflows = synthWorkflows(p);
  expect(workflows[".github/workflows/workflow1.yml"]).toBeDefined();
  expect(workflows[".github/workflows/workflow2.yml"]).toBeDefined();
  expect(workflow3).toBeUndefined();
});

function synthWorkflows(p: Project): any {
  const snapshot = synthSnapshot(p);
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(".github/workflows/"))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {});
  return filtered;
}
