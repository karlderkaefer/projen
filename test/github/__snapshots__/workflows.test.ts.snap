// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding empty workflow 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: my-workflow
on: {}
jobs: {}
"
`;
